File: core/indicators/built_in/rsi.py
"""Relative Strength Index indicator."""

from __future__ import annotations

from typing import Any, Dict

import numpy as np
import pandas as pd

from core.indicators.base_indicator import BaseIndicator


class RSI(BaseIndicator):
    """
    Relative Strength Index (Wilder's smoothing method).

    RSI = 100 - 100 / (1 + RS)
    where RS = avg_gain / avg_loss over ``period`` bars.
    """

    _registry_name = "rsi"

    def __init__(self, period: int = 14, column: str = "close") -> None:
        self.period = period
        self.column = column

    @property
    def name(self) -> str:
        return "rsi"

    @property
    def params(self) -> Dict[str, Any]:
        return {"period": self.period, "column": self.column}

    def calculate(self, df: pd.DataFrame) -> pd.Series:
        """Return the RSI series (0-100)."""
        delta = df[self.column].diff()

        gain = delta.clip(lower=0)
        loss = (-delta).clip(lower=0)

        # Wilder's smoothing (equivalent to EMA with alpha = 1/period).
        avg_gain = gain.ewm(alpha=1.0 / self.period, min_periods=self.period, adjust=False).mean()
        avg_loss = loss.ewm(alpha=1.0 / self.period, min_periods=self.period, adjust=False).mean()

        rs = avg_gain / avg_loss.replace(0, np.nan)
        rsi = 100.0 - (100.0 / (1.0 + rs))

        return rsi

File: apps/web/src/lib/indicators/rsi.ts
import type { OHLCBar } from "@/types";
import { ewm, closes } from "./mathUtils";

export function rsi(
  data: OHLCBar[],
  params: Record<string, unknown>
): (number | null)[] {
  const period = Number(params.period) || 14;
  const close = closes(data);
  const n = close.length;
  const out: (number | null)[] = new Array(n).fill(null);

  if (n < 2) return out;

  // Compute deltas
  const gains: (number | null)[] = [null];
  const losses: (number | null)[] = [null];
  for (let i = 1; i < n; i++) {
    const d = close[i] - close[i - 1];
    gains.push(d > 0 ? d : 0);
    losses.push(d < 0 ? -d : 0);
  }

  // Wilder's smoothing: alpha = 1/period
  const alpha = 1 / period;
  const avgGain = ewm(gains, alpha, period);
  const avgLoss = ewm(losses, alpha, period);

  for (let i = 0; i < n; i++) {
    const ag = avgGain[i];
    const al = avgLoss[i];
    if (ag === null || al === null) continue;
    if (al === 0) {
      out[i] = 100;
    } else {
      out[i] = 100 - 100 / (1 + ag / al);
    }
  }

  return out;
}
